src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+                                     const char*     prologue,
+                                     const char*     prefix,
+                                     int             startIdx,
+                                     int             endIdx,
+                                     const char*     suffix,
+                                     const char*     epilogue);
+static void  appendRangeCalculations(strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  appendLoops            (strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  appendLoopMacroDefs    (strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  appendLoopOuter        (strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  appendLoopInner        (strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  appendLoopMacroUndefs  (strb*           s,
+                                     GEN_KERNEL_CTX* ctx);
+static void  scheduleMaxAndArgmax   (size_t*         blockSize,
+                                     size_t*         gridSize,
+                                     const GpuArray* src,
+                                     const int*      isReduced);
+static void  invokeMaxAndArgmax     (GpuKernel*      kernel,
+                                     const GpuArray* src,
+                                     const int*      isReduced);
+
+
+/* Function implementation */
+
+/**
+ * @brief Computes simultaneously the maxima and the arguments of maxima over
+ * specified axes of the tensor.
+ *
+ * Returns two tensors of identical shape. Both tensors' axes are a subset of
+ * the axes of the original tensor. The axes to be reduced are specified by
+ * the caller, and the maxima and arguments of maxima are computed over them.
+ *
+ * @param [out] dstMax     The resulting tensor of maxima
+ * @param [out] dstArgmax  the resulting tensor of arguments at maxima
+ * @param [in]  src        The source tensor.
+ * @param [in]  isReduced  Either NULL, or an array of booleans of the same
+ *                         size as the dimensionality of the source tensor.
+ *                         Axis k is reduced if isReduced[k] is non-zero,
+ *                         and is preserved otherwise.
+ * @return GA_NO_ERROR if the operation was successful, or a non-zero error
+ *         code otherwise.
+ */
+
+GPUARRAY_PUBLIC int GpuArray_maxandargmax(GpuArray*       dstMax,
+                                          GpuArray*       dstArgmax,
+                                          const GpuArray* src,
+                                          const int*      isReduced){
+	/**
+	 * Generate kernel source code
+	 */
+	
+	const char*  dstMaxType      = gpuarray_get_type(src->typecode) -> cluda_name;
+	const char*  dstArgmaxType   = gpuarray_get_type(GA_LONG)       -> cluda_name;
```

Comment:
@abergeron I thought the strides could be negative, and hence made them signed. Maybe `ssize_t` would be appropriate?

I made everything index-related of type `X` and let it be signed because it's hell to fight C's integer promotion rules: Unsigned outranks signed. 64-bit signed values should store any (unsigned) dimension we could be reasonably interested in for the next few centuries.